samples/torch/src/main/kotlin/Dataset.kt

```diff
@@ -0,0 +1,101 @@
+import kotlinx.cinterop.*
+import platform.posix.*
+
+data class Dataset(val inputs: List<FloatArray>, val labels: List<FloatArray>) {
+    fun batch(indices: List<Int>): Pair<FloatMatrix, FloatMatrix> {
+        val inputBatch = tensor(*(indices.map { inputs[it].toTypedArray() }.toTypedArray()))
+        val labelBatch = tensor(*(indices.map { labels[it].toTypedArray() }.toTypedArray()))
+        return inputBatch to labelBatch
+    }
+
+    fun sampleBatch(batchSize: Int) = batch((0 until batchSize).map { randomInt(inputs.size) })
+    private fun batchAt(batchIndex: Int, batchSize: Int) =
+            batch((0 until inputs.size).drop(batchSize + batchIndex).take(batchSize))
+
+    fun testBatches(batchSize: Int) = (0 until inputs.size / batchSize).map { batchAt(it, batchSize = batchSize) }
+}
+
+/**
+ * Provides the MNIST labeled handwritten digit dataset, described at http://yann.lecun.com/exdb/mnist/
+ */
+class MNIST(val directory: String = "./samples/torch/") {
+    private fun <T> read(fileName: String, action: (Int, CPointer<ByteVar>) -> T): T {
+        val filePath = directory + fileName
+        val file = fopen(fileName, "rb") ?: throw Error("Cannot read input file $filePath")
+
+        try {
+            memScoped {
+                fseek(file, 0, SEEK_END)
+                val fileSize = ftell(file)
+                println("Reading $fileSize bytes from $filePath...")
+                val buffer = allocArray<ByteVar>(fileSize)
```

Comment:
Probably regular ByteArray like in here would be somewhat cleaner: 
https://github.com/JetBrains/kotlinconf-spinner/blob/b9d4550d7611e732f2c54a54a3feeb4d4e1d5f7b/kotlin-native/samples/fullstack/common/src/main/kotlin/kommon/Kommon.kt#L11

Reply:
Great, that made it much better.